### PR TITLE
test: improve code coverage for ZimFileReader.kt

### LINKED ISSUE #4706: Improve code coverage for ZimFileReader.kt
**Particular area and files to be tested**
`ZimFileReader.kt`

**Current Status of the tests**
Incomplete. Needs unit tests to improve coverage as requested in parent issue #4633.

**Particular Behaviour to be tested**

* Execution paths for exception handling (`catch` blocks) during initialization, JNI property getters (`mainPage`, `mediaCount`, `articleCount`), spelling suggestion flows, and asset loading.

Image

### PR DESCRIPTION
**Pull request by harshsomankar123-tech**
Fixes #4706 
- Add ZimFileReaderInstrumentedTest (18 instrumentation tests) covering:
  - Happy paths with real testzim.zim file
  - Exception paths after dispose (exercises catch blocks)
  - Factory.Impl.create with non-existent, corrupted, and valid files

### CODE REVIEW

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+  private suspend fun createZimFileReader(): ZimFileReader {
+    val zimFile = copyZimFile("testzim.zim")
+    val zimReaderSource = ZimReaderSource(zimFile)
+    val archive = zimReaderSource.createArchive()!!
+    return ZimFileReader(zimReaderSource, archive, SuggestionSearcher(archive))
+  }
+
+  @Test
+  fun testMainPageReturnsNonNull() = runBlocking {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: Instead of testing these values in different test cases, we should test these in one test, because it is a UI test, not a unit test. So that the whole ZIM file details should test in one test.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    assertNotNull(reader.mainPage)
+    reader.dispose()
+  }
+
+  @Test
+  fun testMediaCountReturnsValidCount() = runBlocking {
+    val reader = createZimFileReader()
+    val mediaCount = reader.mediaCount
+    assertNotNull(mediaCount)
+    assertTrue(mediaCount!! >= 0)
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: We should test the actual return mediaCount instead of this generic one.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    reader.dispose()
+  }
+
+  @Test
+  fun testArticleCountReturnsValidCount() = runBlocking {
+    val reader = createZimFileReader()
+    val articleCount = reader.articleCount
+    assertNotNull(articleCount)
+    assertTrue(articleCount!! >= 0)
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: This one too. Test the real return `articleCount`.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    reader.dispose()
+  }
+
+  @Test
+  fun testTitleReturnsNonEmpty() = runBlocking {
+    val reader = createZimFileReader()
+    assertTrue(reader.title.isNotEmpty())
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: Instead of testing this non empty we should test the real return title.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    reader.dispose()
+  }
+
+  @Test
+  fun testIdReturnsNonEmpty() = runBlocking {
+    val reader = createZimFileReader()
+    assertTrue(reader.id.isNotEmpty())
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt:108]: Same for id.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+  @Test
+  fun testGetRandomArticleUrlReturnsNonNull() = runBlocking {
+    val reader = createZimFileReader()
+    val url = reader.getRandomArticleUrl()
+    assertNotNull(url)
+    reader.dispose()
+  }
+
+  @Test
+  fun testToBookMapsFieldsCorrectly() = runBlocking {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: From here to the top, all the test cases should be written in one test.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    val reader = createZimFileReader()
+    val book = reader.toBook()
+    assertEquals(reader.title, book.title)
+    assertEquals(reader.id, book.id)
+    assertEquals(reader.creator, book.creator)
+    assertEquals(reader.publisher, book.publisher)
+    assertEquals(reader.date, book.date)
+    assertEquals(reader.language, book.language)
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt:180]: `toBook` has 13 variables; we should test all of them.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+  @Test
+  fun testMainPageReturnsNullAfterDispose() = runBlocking {
+    val reader = createZimFileReader()
+    reader.dispose()
+    assertNull(reader.mainPage)
+  }
+
+  @Test
+  fun testMediaCountReturnsNullAfterDispose() = runBlocking {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: After disposing the `ZIMFileReader` all scenarios should be tested in one test.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    val reader = createZimFileReader()
+    reader.dispose()
+    assertNull(reader.mediaCount)
+  }
+
+  @Test
+  fun testArticleCountReturnsNullAfterDispose() = runBlocking {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: Additionally, you have only tested the main values of `ZimFileReader` but we should test all the fields of it. Like creator, publisher, name, date, description, etc.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+  fun testSearchSuggestionsReturnsNullAfterDispose() = runBlocking {
+    val reader = createZimFileReader()
+    reader.dispose()
+    assertNull(reader.searchSuggestions("test"))
+  }
+
+  @Test
+  fun testGetPageUrlFromReturnsNullForNonExistentTitle() = runBlocking {
+    val reader = createZimFileReader()
+    assertNull(reader.getPageUrlFrom("ThisTitleDefinitelyDoesNotExistInTestZim12345"))
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: Here you have tested, if we provide the wrong title, it returns null for that. But we should also test for the correct title. It will return the path for the correct title.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    reader.dispose()
+  }
+
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt:169]: Test the remaining methods of ZimFileReader, e.g., `getMimeTypeFromUrl`, `isRedirect`, etc, with edge cases.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+  @Test
+  fun testFactoryCreateReturnsNullForNonExistentFile() = runBlocking {
+    val factory = ZimFileReader.Factory.Impl()
+    val nonExistentSource = ZimReaderSource(File("/non/existent/path/fakefile.zim"))
+    val result = factory.create(nonExistentSource, false)
+    assertNull(result)
+  }
+
+  @Test
+  fun testFactoryCreateReturnsNullForCorruptedFile() = runBlocking {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: Good test, appreciate it 👏 .

**Comment by harshsomankar123-tech** [comment]
@MohitMaliFtechiz PTAL

**Comment by kelson42** [comment]
Sorry, I have rebased. I though PR is ready to merge.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    val zimReaderSource = ZimReaderSource(zimFile)
+    val archive = zimReaderSource.createArchive()!!
+    return ZimFileReader(zimReaderSource, archive, SuggestionSearcher(archive))
+  }
+
+  @Test
+  fun testZimFileReaderProperties() = runBlocking {
+    val reader = createZimFileReader()
+
+    assertNotNull(reader.mainPage)
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: @harshsomankar123-tech We should verify the values come from the reader instead of just checking the they are not null.  Because the ZIM file we are using in this, we know all the details of it.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+    assertNull(reader.articleCount)
+    assertNull(reader.searchSuggestions("test"))
+    assertNull(reader.getRandomArticleUrl())
+    assertEquals("No Title Found", reader.title)
+    assertEquals("", reader.creator)
+    assertEquals("", reader.publisher)
+    assertEquals("", reader.date)
+    assertEquals("", reader.description)
+    assertEquals("", reader.language)
+    assertEquals("", reader.tags)
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/core/reader/ZimFileReaderInstrumentedTest.kt]: We should test the remaining values as well. There are 14 variables in `reader`, so we should test those in both scenarios. When the reader is created with a ZIM file, and when the reader is disposed.

**Comment by harshsomankar123-tech** [comment]
@MohitMaliFtechiz Thanks! PTAL
